### PR TITLE
feat(baselines): Add SSFL baseline

### LINKED ISSUE #7811: [Baseline] Add SSFL (Salient Sparse Federated Learning)
### Type

New Baseline

### Description

I would like to contribute a Flower Baseline for **SSFL: Discovering Unified Sparse Subnetworks at Initialization for Efficient Federated Learning**.

- **Paper (TMLR / OpenReview):** https://openreview.net/forum?id=kUZ6LhUB26
- **Reference implementation:** https://github.com/riohib/SSFL

### Method
SSFL (Salient Sparse Federated Learning) discovers a unified sparse subnetwork at initialization by aggregating client-local saliency scores. Clients then train within this fixed shared subspace and communicate only active parameters. The paper provides early evidence that learning in a unified sparse subspace can improve the accuracy–sparsity trade-off compared with dynamic pruning methods that continually change the sparse topology during training, while reducing communication relative to dense FL.

#### Why this is a useful Flower baseline
- Reproduces a published method with a clear, self-contained pipeline .
- Targets a common research need: sparse / communication-efficient FL with non-IID data.
- Fits Flower’s stack well (PyTorch, Flower Datasets, Message API, sparse payload transport).
- Provides paper-matching CIFAR-10/100 configs that others can extend or compare against.

### Planned experiments
- CIFAR-10 and CIFAR-100
- ResNet-18
- Static SSFL at density `0.5`
- 100 clients, 10% participation per round
- Balanced Dirichlet partitioning with `α = 0.3`
- Paper-matching optimizer and training configurations for CIFAR-10 and CIFAR-100

### Planned Implementation
1. Port the our Flower App into `baselines/ssfl`.
2. Reproduce SSFL’s saliency-based mask discovery and sparse training using Flower Datasets and the Message API.
3. Add paper configurations and documentation for CIFAR-10/100.

### Additional Context
I am one of the paper’s authors.

A working Flower App port and paper-scale runs already exist and we are testing to share soon. We plan to contribute the CIFAR-10/100 static SSFL baseline first, then potentially add more experiments later.

The baseline will use **PyTorch**, **Flower Datasets**, and the **Flower Message API**.

### PR DESCRIPTION
**Pull request by riohib**
### Description

This PR contributes a Flower Baseline for **SSFL: Discovering Unified Sparse Subnetworks at Initialization for Efficient Federated Learning**.

SSFL discovers a unified sparse subnetwork at initialization by aggregating client-local saliency scores. Clients then train within this fixed shared subspace and communicate only the active parameters. The paper demonstrates that learning in a unified sparse subspace can improve the accuracy–sparsity trade-off compared with methods that continually change the sparse topology.

- Paper: https://openreview.net/forum?id=kUZ6LhUB26
- Reference implementation: https://github.com/riohib/SSFL

### Related issues/PRs

Closes #7811

### Explanation

This PR adds `baselines/ssfl` with:

- A Flower Message API implementation of static SSFL
- Client-side saliency computation and server-side global mask aggregation
- Sample-weighted sparse FedAvg
- Flower Datasets integration with balanced Dirichlet partitioning
- ResNet-18 experiments for CIFAR-10 and CIFAR-100
- Paper-scale configurations using 100 clients and 10% participation
- Offline W&B logging, checkpoints, JSONL metrics, and communication accounting
- Documentation covering setup, experiment commands, expected results, and scope

Reference single-seed results (`seed=550`):

| Dataset | Flower baseline | Paper 5-run mean |
| --- | ---: | ---: |
| CIFAR-10 | 89.56% | ~88.29% |
| CIFAR-100 | 59.75% | ~61.37% |

The dependency currently targets Flower 1.30 because Flower 1.31+ and `flwr-datasets` 0.6 have incompatible `rich` version constraints.

Validation completed locally:

- isort
- Black
- Ruff
- mypy
- Pylint (`10.00/10`)
- pytest (`27 passed`)
- Standard `pip install -e .`

### Checklist

- [x] Implement proposed change
- [x] Write tests
- [x] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [x] Address LLM-reviewer comments, if applicable (e.g., GitHub Copilot)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### Any other comments?

This initial contribution focuses on SSFL with ResNet-18 on CIFAR-10 and CIFAR-100 at density `0.5`.

I am one of the SSFL paper’s authors and am happy to address review feedback or add further reproduction details.

### CODE REVIEW

File: baselines/ssfl/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+[project.optional-dependencies]
+dev = [
+    "isort==5.13.2",
+    "black==24.2.0",
+    "mypy==1.8.0",
+    "pylint==3.2.6",
+    "flake8==5.0.4",
+    "pytest>=8.0.0",
+    "ruff==0.1.9",
+]
```
[the remark this answers is not in the document]
Comment (riohib) [reply baselines/ssfl/pyproject.toml:36]: Left wandb as a main dependency so flwr run installs it into the isolated runtime env; logging stays opt-in via wandb-mode.

File: baselines/ssfl/ssfl/model.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+        pruned_load_dict = OrderedDict()
+        for key, value in state_dict.items():
+            base_name = key.rsplit(".", 1)[0]
+            if key.endswith(".weight") and base_name in self.pruned_layers:
+                pruned_load_dict[f"{base_name}.weight_orig"] = value
+            else:
+                pruned_load_dict[key] = value
+        return self._wrapped_model().load_state_dict(
+            pruned_load_dict, strict=False, assign=assign
+        )
```
[the remark this answers is not in the document]
Comment (riohib) [reply baselines/ssfl/ssfl/model.py]: Addressed in 68154615b — SparseModel.load_state_dict now checks the public flattened keys and raises on missing/unexpected keys when strict=True.

File: baselines/ssfl/tests/test_data_path.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+
+@pytest.fixture(scope="module")
+def cifar_root(tmp_path_factory) -> Path:
+    # Prefer a persistent local-disk cache; NFS temp dirs make torchvision
+    # downloads extremely slow on shared cluster hosts.
+    env_root = os.environ.get("SSFL_TEST_CIFAR_ROOT", "").strip()
+    if env_root:
+        root = Path(env_root)
+    else:
```
[the remark this answers is not in the document]
Comment (riohib) [reply baselines/ssfl/tests/test_data_path.py]: Addressed in 68154615b — test_data_path.py now skips unless SSFL_TEST_CIFAR_ROOT is set.

File: baselines/ssfl/ssfl/server_app.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+    )
+    torch.save(arrays.to_torch_state_dict(), directory / f"{stem}_model.pt")
+    torch.save(masks_to_cpu_uint8(masks), directory / f"{stem}_mask.pt")
+    (directory / f"{stem}_mask_version.txt").write_text(digest + "\n")
+
+
+def _append_jsonl(path: Path, record: dict[str, Any]) -> None:
+    path.parent.mkdir(parents=True, exist_ok=True)
+    with path.open("a", encoding="utf-8") as handle:
+        handle.write(json.dumps(record, sort_keys=True) + "\n")
```
[the remark this answers is not in the document]
Comment (riohib) [reply baselines/ssfl/ssfl/server_app.py:218]: Addressed in 68154615b — the server now replaces metrics.jsonl at the start of each run so reruns do not mix event

File: baselines/ssfl/ssfl/mask.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    Create a binary mask by keeping scores >= the k-th largest score.
+
+    Ties at the threshold are retained (legacy SSFL behavior), so the resulting
+    density can slightly exceed keep_ratio.
+    """
+    if not scores_dict:
+        raise ValueError("The saliency score dictionary is empty.")
+
+    all_scores = torch.cat([v.flatten() for v in scores_dict.values()])
+    num_params_to_keep = int(len(all_scores) * keep_ratio)
```
[the remark this answers is not in the document]
Comment (riohib) [reply baselines/ssfl/ssfl/mask.py:78]: Addressed in 68154615b — create_mask_from_scores now raises if keep_ratio is outside [0, 1].

File: baselines/ssfl/ssfl/comm_stats.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+
+@dataclass
+class CommStats:
+    """Byte/param counters. Payload bytes are ArrayRecord payloads, not wire bytes."""
+
+    discovery_downlink_payload_bytes: int = 0
+    discovery_uplink_payload_bytes: int = 0
+    mask_downlink_payload_bytes: int = 0
+    train_uplink_payload_bytes: int = 0
```
[the remark this answers is not in the document]
Comment (riohib) [reply baselines/ssfl/ssfl/comm_stats.py:16]: Addressed in e5f75230b — training downlink is now arrays_to_send.count_bytes() × sampled clients each round and is included in CommStats / summary.json.

File: baselines/ssfl/ssfl/server_app.py [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+    mask_downlink = _install_masks(
+        grid, masks=masks, digest=digest, node_ids=node_ids, timeout=timeout
+    )
+    comm.mask_downlink_payload_bytes = mask_downlink
+    mask_install_metrics = {
+        "event": "mask_installed",
+        "server_round": 0,
+        "comm/mask_downlink_payload_bytes": mask_downlink,
+    }
+    wandb_session.log(mask_install_metrics, step=0)
```
[the remark this answers is not in the document]
Comment (riohib) [reply baselines/ssfl/ssfl/server_app.py]: Addressed in bcffd2f0f — step 0 now combines init metrics and stays open until the first evaluation so W&B does not drop mask-install or the round-zero eval.

File: baselines/ssfl/ssfl/server_app.py [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+                ),
+                **{
+                    f"train/{key}": (
+                        float(value) if isinstance(value, (int, float)) else value
+                    )
+                    for key, value in dict(round_metrics).items()
+                },
+            }
+            metrics_history.append(train_record)
+            _append_jsonl(checkpoint_dir / "metrics.jsonl", train_record)
```
[the remark this answers is not in the document]
Comment (riohib) [reply baselines/ssfl/ssfl/server_app.py]: Addressed in 019be3949 — each train record is now logged to W&B as the round completes, independent of save-metrics, and the step stays open when centralized eval will use the same round.

File: baselines/ssfl/ssfl/server_app.py [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+
+    strategy = SSFLStrategy(
+        node_to_client_id=node_to_client,
+        masks=masks,
+        mask_version=digest,
+        sample_seed=seed,
+        transport=transport,
+        on_train_round=_on_train_round,
+        fraction_train=float(cfg["fraction-train"]),
+        fraction_evaluate=float(cfg["fraction-evaluate"]),
```
[the remark this answers is not in the document]
Comment (riohib) [reply baselines/ssfl/ssfl/server_app.py:406]: Intentional — all profiles use fraction-evaluate = 0.0 and centralized server eval; client-side evaluate is out of scope for this baseline.

File: baselines/ssfl/ssfl/server_app.py [hunk 10]
```diff
@@ -0,0 +1,529 @@
...
+def _metric_float(value: object) -> float:
+    if not isinstance(value, (int, float)):
+        raise TypeError("Expected a scalar numeric metric")
+    return float(value)
+
+
+def _device(prefer_cpu: bool = True) -> torch.device:
+    if prefer_cpu or not torch.cuda.is_available():
+        return torch.device("cpu")
+    return torch.device("cuda:0")
```
Comment (jafermarq) [inline baselines/ssfl/ssfl/server_app.py:66]: would it be possible to move these to a shared, for example,`app_utils.py` module that can be used by `client_app.py` and `server_app.py`. There's only a bit of overlap in functionality but it would help keeping the `server_app.py` and `client_app.py` more "focused" in the functionality. wdyt?

Possibly some of the helper functions below this comment could be moved too. Up to you

File: baselines/ssfl/ssfl/partitioner.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+from __future__ import annotations
+
+from typing import Any, cast
+
+import numpy as np
+from datasets import Dataset
+from flwr_datasets.partitioner import Partitioner
+
+
+class BalancedDirichletPartitioner(Partitioner):
```
Comment (jafermarq) [inline baselines/ssfl/ssfl/partitioner.py:12]: Nice! Would you me interested in adding this partitioner to `flwr-datasets`? We'll are getting the _conflict_ with `rich` fixed soon so we are planning an new `flwr-datasetes` release.

File: baselines/ssfl/pyproject.toml [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+[project]
+name = "ssfl"
+version = "1.0.0"
+description = "SSFL: Discovering Unified Sparse Subnetworks at Initialization for Efficient Federated Learning"
+license = { file = "LICENSE" }
+requires-python = ">=3.12"
+dependencies = [
+    # Flower 1.31+ and flwr-datasets 0.6.0 have incompatible rich bounds.
+    # Flower 1.30 provides the Message API used here and resolves with pip.
+    "flwr[simulation]>=1.28.0,<1.31.0",
```
Comment (jafermarq) [inline baselines/ssfl/pyproject.toml:14]:
```suggestion
    "flwr[simulation]>=1.28.0",
```

You are right about `rich`. I think we can leave this unbounded so it runs on more recent version of `flwr` when they become compatible with `flwr-datasets`. Let's leave this comment open as we are working on a new release of `flwr-dataset` that will lift this limitation

File: baselines/ssfl/ssfl/partitioner.py (hunk 11 above)
Comment (jafermarq) [reply baselines/ssfl/ssfl/partitioner.py:12]: We just released `flwr-datasets 0.6.1`

File: baselines/ssfl/pyproject.toml (hunk 12 above)
Comment (jafermarq) [reply baselines/ssfl/pyproject.toml:14]: We just released `flwr-datasets 0.6.1`
